packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+  tags?: string | TagMap | (string | TagMap)[];
+};
+
+/**
+ * Decorate a class with binding configuration
+ *
+ * @example
+ * ```ts
+ * @bind((binding) => {binding.inScope(BindingScope.SINGLETON).tag('controller')}
+ * )
+ * export class MyController {
+ * }
+ * ```
+ *
+ * @param templates A list of binding scope/tags or template functions to
+ * configure the binding
+ */
+export function bind(
+  ...templates: (BindingTemplate | BindingScopeAndTags)[]
+): ClassDecorator {
+  const templateFunctions = templates.map(t => {
+    if (typeof t === 'function') {
+      return t;
+    } else {
+      return asBindingTemplate(t);
+    }
+  });
+
+  return (target: Function) => {
+    const spec: BindingMetadata = {
+      templates: [
+        // Add a template to bind to a class
+        binding => {
+          binding.toClass(target as Constructor<unknown>);
+        },
+        ...templateFunctions,
+      ],
+    };
+
+    const decorator = BindDecoratorFactory.createDecorator(
+      BINDING_METADATA_KEY,
+      spec,
+    );
+    decorator(target);
+  };
+}
+
+export namespace bind {
+  /**
+   * `@bind.provider` to denote a provider class
+   *
+   * A list of binding scope/tags or template functions to configure the binding
+   */
+  export function provider(
+    ...templates: (BindingTemplate | BindingScopeAndTags)[]
+  ): ((target: Constructor<Provider<unknown>>) => void) {
+    return (target: Constructor<Provider<unknown>>) => {
+      bind(
+        // Set up the default for providers
+        binding =>
+          binding.toProvider(target).tag('provider', {type: 'provider'}),
+        // Call other template functions
+        ...templates,
+      )(target);
+    };
+  }
+}
+
+/**
+ * Convert binding scope and tags as a template function
+ * @param scopeAndTags
+ */
+function asBindingTemplate(scopeAndTags: BindingScopeAndTags): BindingTemplate {
+  return binding => {
+    if (scopeAndTags.scope) {
+      binding.inScope(scopeAndTags.scope);
+    }
+    if (scopeAndTags.tags) {
+      if (Array.isArray(scopeAndTags.tags)) {
+        binding.tag(...scopeAndTags.tags);
+      } else {
+        binding.tag(scopeAndTags.tags);
+      }
+    }
+  };
+}
+
+/**
+ * Get binding metadata for a class
+ * @param target The target class
+ */
+export function getBindingMetadata(
+  target: Function,
+): BindingMetadata | undefined {
+  return MetadataInspector.getClassMetadata<BindingMetadata>(
+    BINDING_METADATA_KEY,
+    target,
+  );
+}
+
+/**
+ * Get the binding template from a class with binding metadata
+ *
+ * @param cls A class with optional `@bind`
+ */
+export function bindingTemplateFor(cls: Constructor<unknown>): BindingTemplate {
+  const spec = getBindingMetadata(cls);
+  const templateFunctions = (spec && spec.templates) || [];
+  return binding => {
+    for (const t of templateFunctions) {
+      binding.apply(t);
+    }
+    if (spec && spec.target !== cls) {
+      // Remove name/key from base classes
```

Comment:
This comment can be read in two ways:
 1. modify base classes and remove name/key
 2. modify binding and remove name/key values set by base classes

I think the second option is describing your intent, in which case I am proposing slightly different wording:

```
Remove name & key tags that were configured by base classes
```